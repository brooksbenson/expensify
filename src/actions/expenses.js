import db from '../firebase/firebase';
      
export function addExpense(expense) {
  return {
    type: 'ADD_EXPENSE',
    expense 
  };
}

export function startAddExpense(expenseData = {}) {
  const { 
    description = '', 
    note = '', 
    amount = 0, 
    createdAt = new Date().valueOf() 
  } = expenseData;

  return async (dispatch) => {
    const expense = { description, note, amount, createdAt };
    const ref = await db.ref('expenses').push(expense);
    dispatch(addExpense({
      id: ref.key,
      ...expense
    }));
  }
}

export function removeExpense(id) {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
}

export function startRemoveExpense(id) {
  return async (dispatch) => {
    await db.ref(`expenses/${id}`).remove();
    dispatch(removeExpense(id));
  }
}

export function editExpense(id, update) {
  return {
    type: 'EDIT_EXPENSE',
    id,
    update
  }
}

export function startEditExpense(id, update) {
  return async (dispatch) => {
    await db.ref(`expenses/${id}`).update({ ...update });
    dispatch(editExpense(id, update));
  }
}

export function setExpenses(expenses) {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
}

export function startSetExpenses() {
  return async (dispatch) => {
    const snapshot = await db.ref('expenses').once('value');
    const expenses = [];
    snapshot.forEach(child => {
      expenses.push({
        id: child.key,
        ...child.val()
      });
    });
    dispatch(setExpenses(expenses));
  }
}