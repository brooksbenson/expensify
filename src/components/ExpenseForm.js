import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    
    const e = props.expense;
    this.state = {
      description: e ? e.description : '',
      amount: e ? (e.amount / 100).toString() : '',
      note: e ? e.note : '',
      createdAt: e ? moment(e.createdAt) : moment(),
      calFocused: false,
      error: false
    }
  }

  onDescriptionChange = (e) => {
    const {value} = e.target;
    this.setState(() => ({description: value}))
  }

  onAmountChange = (e) => {
    const {value} = e.target;
    const validator = /^\d*(\.\d{0,2})?$/;
    if (!value || validator.test(value)) {
      this.setState(() => ({amount: value}))
    }
  }

  onNoteChange = (e) => {
    const {value} = e.target;
    this.setState(() => ({note: value}))
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}));
    }
  }

  onCalFocusChange = ({focused}) => {
    this.setState(() => ({calFocused: focused}));
  }

  onSubmit = (e) => {
    e.preventDefault();
    const {description, amount, note, createdAt} = this.state;
    if (!description || !amount) {
      this.setState(() => ({error: true}));
    } else {
      this.setState(() => ({error: false}));
      this.props.onSubmit({
        description,
        note,
        amount: Math.round(Number(amount) * 100),
        createdAt: createdAt.valueOf()
      })
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error"> Please provide an amount and description </p>}
        <input
          type="text"
          placeholder="description"
          className="text-input"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        >
        </input>
        <input
          type="text"
          placeholder="amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        >
        </input>
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calFocused}
          onFocusChange={this.onCalFocusChange}
          numberOfMonths={1}
          isOutsideRange={day => false}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <button className='button'>
          {this.props.expense ? 'Save Expense' : 'Add Expense' }
        </button>
      </form> 
    );
  }
}