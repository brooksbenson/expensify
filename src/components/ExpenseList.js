import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

export function ExpenseList({expenses}) {
  return (
    <div>
      {
        expenses.length === 0 
          ? <p> No expenses </p>
          : expenses.map(e => <ExpenseListItem key={e.id} {...e} />)
      }
    </div>
  );
}

function mapStateToProps({ expenses, filters }) {
  return {expenses: selectExpenses(expenses, filters)};
}
export default connect(mapStateToProps)(ExpenseList);