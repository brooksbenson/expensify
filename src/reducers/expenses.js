const expensesDefaultState = [];
export default function(state = expensesDefaultState, action) {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(e => e.id != action.id);
    case 'EDIT_EXPENSE':
      return state.map(e => e.id === action.id ? {...e, ...action.update} : e);
    default:
      return state;
  }
}