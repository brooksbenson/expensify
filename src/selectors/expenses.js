import moment from 'moment';

export default function(expenses, { text, startDate, endDate, sortBy }) {
  const search = new RegExp(text, 'i');
  return expenses.filter(e => {
    const createdAt = moment(e.createdAt);
    const textMatch = search.test(e.description) || search.test(e.note);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt, 'day') : true;

    return textMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') return a.createdAt < b.createdAt ? 1 : -1;
    if (sortBy === 'amount') return a.amount < b.amount ? 1 : -1;
  });
}