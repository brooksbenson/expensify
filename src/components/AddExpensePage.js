import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from './../actions/expenses';

export function AddExpensePage({addExpense, history}) {
  return (
    <div>
      <ExpenseForm 
        onSubmit={e => {
          addExpense(e)
          history.push('/')
        }}
      />
    </div>
  );
};

function mapDispatchToProps(dispatch) {
  return {
    addExpense(e) {
      dispatch(addExpense(e));
    }
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);