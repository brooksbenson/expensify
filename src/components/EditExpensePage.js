import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from './../actions/expenses';

export function EditExpensePage({expense, history, editExpense, removeExpense}) {
  return (
    <div>
      <ExpenseForm
        expense={expense} 
        onSubmit={update => {
          editExpense(expense.id, update);
          history.push('/');
        }}
      />
      <button onClick={() => {
        removeExpense(expense.id)
        history.push('/');
      }}>
        remove
      </button>
    </div>
  );
}

function mapDispatchToProps(dispatch) {
  return {
    editExpense(id, update) {
      dispatch(editExpense(id, update));
    },
    removeExpense(id) {
      dispatch(removeExpense(id));
    }
  }
}

function mapStateToProps({expenses}, {match}) {
  const {id} = match.params;
  return {
    expense: {...expenses.find(e => e.id === id)}
  } 
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);