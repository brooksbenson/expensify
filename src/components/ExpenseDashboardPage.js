import React from 'react';
import { connect } from 'react-redux';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseList from './ExpenseList';
import { updateFilters } from '../actions/filters';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseDashboardPage = ({expenses, expensesCount, expensesTotal, filters, updateFilters}) => (
  <div>
    <ExpensesSummary expensesTotal={expensesTotal} expensesCount={expensesCount} />
    <ExpenseListFilters {...filters} updateFilters={updateFilters} />
    <ExpenseList expenses={expenses} />
  </div>
);

const mapStateToProps = ({expenses, filters}) => {
  const selectedExpenses = selectExpenses(expenses, filters);
  return {
    expenses: selectedExpenses,
    expensesTotal: selectExpensesTotal(selectedExpenses),
    expensesCount: selectedExpenses.length,
    filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  updateFilters(update) {
    dispatch(updateFilters(update));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseDashboardPage);