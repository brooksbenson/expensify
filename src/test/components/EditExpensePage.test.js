import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let wrapper, editExpense, removeExpense, history, expense;
beforeEach(() => {
  [expense] = expenses;
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = {push: jest.fn()};
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expense}
    />
  );
});

test('should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should respond correctly to edit expense', () => {
  let update = expenses[1];
  delete update.id;
  wrapper.find('ExpenseForm').prop('onSubmit')(update);
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, update);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should respond correctly to delete expense', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenLastCalledWith(expense.id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});