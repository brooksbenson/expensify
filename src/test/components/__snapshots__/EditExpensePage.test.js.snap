// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1000,
        "createdAt": 604800000,
        "description": "car wash",
        "id": "1",
        "note": "clean car",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
